cdap-ui/app/cdap/components/LogViewer/index.tsx

```diff
@@ -0,0 +1,387 @@
+    logsContainer.addEventListener('scroll', this.watchScrollCallback);
+  };
+
+  private stopScrollWatch = () => {
+    const logsContainer = this.logsContainer.current;
+    logsContainer.removeEventListener('scroll', this.watchScrollCallback);
+  };
+
+  private scrollToBottom = () => {
+    const logsContainer = this.logsContainer.current;
+    logsContainer.scrollTop = logsContainer.scrollHeight;
+    this.scrollPosition = logsContainer.scrollTop;
+  };
+
+  private startPoll = () => {
+    if (!this.state.isPolling) {
+      this.setState({ isPolling: true });
+    }
+
+    this.props.dataFetcher.getNext().subscribe((res) => {
+      if (res.length > 0) {
+        const newLogs = this.state.logs.concat(res);
+        this.setState({ logs: newLogs });
+      }
+
+      this.setState({ isFetching: false }, () => {
+        this.scrollToBottom();
+        this.trimTopExcessLogs();
+
+        this.pollTimeout = setTimeout(this.startPoll, 5000);
+      });
+    });
+  };
+
+  private stopPoll = () => {
+    if (this.pollTimeout) {
+      clearTimeout(this.pollTimeout);
+      this.pollTimeout = null;
+    }
+
+    if (this.state.isPolling) {
+      this.setState({ isPolling: false });
+    }
+  };
+
+  private fetchPrev = debounce(() => {
+    if (this.state.isFetching || this.state.isPolling) {
+      return;
+    }
+    this.setState({ isFetching: true });
+
+    const logsContainer = this.logsContainer.current;
+    const currentScrollHeight = logsContainer.scrollHeight;
+
+    this.props.dataFetcher.getPrev().subscribe(
+      (res) => {
+        if (res.length > 0) {
+          const newLogs = res.concat(this.state.logs);
+          this.setState({ logs: newLogs });
+        }
+
+        this.setState({ isFetching: false }, () => {
+          // maintaining scroll position
+          const updatedScrollHeight = logsContainer.scrollHeight;
+          logsContainer.scrollTop = updatedScrollHeight - currentScrollHeight;
+
+          const trimmedLogs = this.state.logs.slice(0, MAX_LOG_ROWS);
+
+          // trim excess logs
+          this.setState({
+            logs: trimmedLogs,
+          });
+          this.props.dataFetcher.onLogsTrim(
+            trimmedLogs[0],
+            trimmedLogs[this.state.logs.length - 1]
+          );
+        });
+      },
+      (err) => {
+        // tslint:disable-next-line: no-console
+        console.log('err', err);
+      }
+    );
+  }, 100);
+
+  private fetchNext = debounce(() => {
+    if (this.state.isFetching || this.state.isPolling) {
+      return;
+    }
+    this.setState({ isFetching: true });
+
+    this.props.dataFetcher.getNext().subscribe(
+      (res) => {
+        if (res.length > 0) {
+          const newLogs = this.state.logs.concat(res);
+          this.setState({ logs: newLogs });
+        }
+
+        this.setState({ isFetching: false }, this.trimTopExcessLogs);
+      },
+      (err) => {
+        // tslint:disable-next-line: no-console
+        console.log('err', err);
```

Comment:
nit: We should start surfacing the errors if the API to fetch logs fails.. We generally log in console and to debug later but lets surface this up so that its self serving if the user doesn't see any logs.